### ISSUE TITLE
api: move `polars.from_dataframe` to `polars.interchange.from_dataframe`, or even consider removing?

### ISSUE THREAD

**Issue by MarcoGorelli** [pola-rs/polars#20065]
`polars.from_dataframe` uses the dataframe interchange protocol to convert between dataframes. However, users aren't necessarily are of this, nor are they aware of the limitations of the interchange protocol. Proof: at EuroScipy this year I saw someone use `pl.from_dataframe` to convert from pandas to Polars because they thought that was just the recommended way of doing it

The reason this matters is that the interchange protocol is tied down to pandas/polars implementations. And pandas implementation had some severe (critical?) bugs before 2.2, which users wouldn't necessarily be aware of. A new Polars user could well use `pl.from_dataframe` with a pandas dataframe, get nonsense data back, and blame Polars even though the bug was on the pandas side

So, my first inclination would be to move `polars.from_dataframe` to `polars.interchange.from_dataframe`

But second, we should probably have a conversation about whether we should keep it at all? Polars already supports the PyCapsule Interface now for both import and export, so anyone wishing to agnostically convert between dataframe can already use that, and with that it also opens the doors to agnostically accessing the underlying data from say C or Rust

In terms of ecosystem:
- the interchange protocol hasn't been updated since August 2023. It still doesn't support Date, Duration, or nested dtypes, and I can't see that changing, there's radio silence in the repo. This is a bit sad to me - we had our disagreements over the "dataframe api standard" project, but I was hoping that at least the interchange protocol would have continued
- All open source users of the interchange protocol have switched to either the PyCapsule Interface or they've made their code totally dataframe agnostic via Narwhals. The only one left is Seaborn, and they're debating what to do, but they don't seem keen on keeping the interchange protocol around
- cuDF have announced their intention to deprecate support for it https://github.com/rapidsai/cudf/issues/17403

What I feel bad about is that Stijn was the only person to have read the interchange protocol spec carefully enough to have come up with a correct and useful implementation, and it would be a pity to see that effort go to waste. Do we take ownership of the interchange protocol and drive it forwards, or just let it sink and encourage the PyCapsule Interface for the same use cases?

**TL;DR**:
- the dataframe interchange protocol has some design and implementation issues that don't  look like they'll ever get resolved
- The PyCapsule Interface allows for the same things, but much more robustly
- cuDF want to remove support for the interchange protocol
- downstream libraries have moved to different solutions

**Comment by MarcoGorelli** [comment]
Initial response from discussion:
- OK to move `polars.from_dataframe` to `polars.interchange.from_dataframe`
- there's no hurry to throw out support for the protocol immediately, we can keep monitoring the situation and see how things progress

**Comment by MarcoGorelli** [comment]
update: cuDF have indeed deprecated `__dataframe__`

In pandas, `from_dataframe` now gives priority to the Arrow PyCapsule Interface: https://github.com/pandas-dev/pandas/pull/60739/files. I think the same could be done here in Polars, that may be less disruptive than removing / deprecating

**Comment by mcrumiller** [comment]
@MarcoGorelli this issue is referenced in the [`pytest` section of `pyproject.toml`](https://github.com/pola-rs/polars/blob/main/py-polars/pyproject.toml#L289C11-L289C57). It's been over a year, is there an update on it?

**Comment by MarcoGorelli** [comment]
hey - yeah it's on my todo list to deprecate it (it's now deprecated in pandas too), just haven't got round it. feel free to deprecate it if you have time/interest, else i'll get to it before too long hopefully
